i18n_scripts/translation.py

```diff
@@ -0,0 +1,364 @@
+        languages_dirs = [
+            directory for directory in os.listdir(lang_parent_dir)
+            if (
+                    directory.startswith('values-')
+                    and 'strings.xml' in os.listdir(os.path.join(lang_parent_dir, directory))
+            )
+        ]
+        return languages_dirs
+    except FileNotFoundError as e:
+        print(f"Directory not found: {e}", file=sys.stderr)
+        raise
+    except PermissionError as e:
+        print(f"Permission denied: {e}", file=sys.stderr)
+        raise
+
+
+def separate_translation_to_modules(modules_dir, lang_dir):
+    """
+    Separates translations from a translation file into modules.
+
+    Args:
+        modules_dir (str): The directory containing all the modules.
+        lang_dir (str): The directory containing the translation file being split.
+
+    Returns:
+        dict: A dictionary containing the translations separated by module.
+        {
+            'module_1_name': etree.Element('resources')_1.
+            'module_2_name': etree.Element('resources')_2.
+            ...
+        }
+    """
+    translations_roots = {}
+    try:
+        # Parse the translation file
+        file_path = get_translation_file_path(modules_dir, 'i18n', lang_dir)
+        module_translations_tree = etree.parse(file_path)
+        root = module_translations_tree.getroot()
+        previous_entry = None
+        # Iterate through translation entries
+        for translation_entry in root.getchildren():
+            if not isinstance(translation_entry, etree._Comment):
+                # Split the key to extract the module name
+                module_name, key_remainder = translation_entry.attrib['name'].split('.', maxsplit=1)
+                translation_entry.attrib['name'] = key_remainder
+
+                # Create a dictionary entry for the module if it doesn't exist
+                if module_name not in translations_roots:
+                    translations_roots[module_name] = etree.Element('resources')
+                    translations_roots[module_name].text = '\n\t'
+
+                # Append the translation entry to the corresponding module
+                append_element_and_comment(translation_entry, previous_entry, translations_roots[module_name])
+
+            previous_entry = translation_entry
+        return translations_roots
+
+    except FileNotFoundError as e:
+        print(f"Error: Translation file not found: {e}", file=sys.stderr)
+        return translations_roots  # Return an empty dictionary if file not found
+
+    except etree.XMLSyntaxError as e:
+        print(f"Error: XML syntax error in translation file: {e}", file=sys.stderr)
+        return translations_roots  # Return an empty dictionary if XML syntax error occurs
+
+    except Exception as e:
+        print(f"Error: An unexpected error occurred: {e}", file=sys.stderr)
+        return translations_roots  # Return an empty dictionary for any other unexpected error
```

Comment:
@Amr-Nash Do you think this is a valid way to handle the error? I think we should `print` a message and `raise` the error.

Reply:
I tend to align with @OmarIthawi's perspective. While the handling of the first exception seems acceptable, it's imperative to re-raise the last two exceptions. Additionally, the try-catch blocks appear to encompass a significant portion of code. I suggest breaking down this function into distinct try-catch blocks. For instance, one for file retrieval and another for loop continuation, which would enhance clarity and alleviate confusion regarding the origin of errors and which-line-throws-what-exception.

Reply:
Thank you @OmarIthawi and @iamjazzar for your feedback.
I refactored some of the error handling and added some inside the for loops you can recheck them when you have time